# MPA 5830 - Module 02
In this module our goal is to understand how some very powerful packages work their magic. In particular, although they are of somewhat recent vintage, `{dplyr}`, `{tidyr}` and `{lubridate}` have quickly gained a fan following because they allow you to clean and organize your data and then calculate quantities of interest with surprising ease. In this module we will start to see some of these packages' core functionalities and wrap up this particular leg of our learning journey in __`Module 03`__. 

## `{dplyr}` 
There is a common quote that is tossed about a good bit in the hallways of data science, that, and I am paraphrasing here, a data scientists spends about 80% of their time gathering, cleaning, and organizing their data and spends only about 20% of their time on the analysis per se. This may or may not be true, especially in the initial stages of a new project but yes, we do spend an awfully large amount of our time getting the data ready for visualizations and other analyses. You do this work long enough and you come to realize that anything you could do to speed up the cleaning phase would be time and money saved. And yet, data cleaning skills are in short supply. On the plus side of the ledger, packages like `{dplyr}` and `{data.table}` have simplified what were once nightmarish tasks. 

`Nightmare` is not a word to be tossed around lightly and so let us setup a seemingly large data-set with 100+ columns, and tons of information hidden in it. Once we setup this scenario, spell out a few questions we would like to answer, we might better appreciate how `{dplyr}` comes to our aid. In particular, we might come to understand how `{dplyr}` uses seven core verbs:  

| **What you want to do ...** | **`{dplyr}` function** |
| :-----  | :----- |
| you need to select columns to work with? | `select()`|
| you need to use a subset of the data based on some criterion? | `filter()`|
| you need to arrange the data in ascending/descending order of variable(s)? | `arrange()`|
| you want the results of your calculations to be a standalone data frame? | `summarise()`|
| you want to add your calculated value(s) to the existing data frame? | `mutate()`|
| you want to add your calculated value(s) to the existing data frame but also  drop all  variables/columns not  being used in the calculation? | `transmute()`|
| you need to calculate averages, frequencies, etc by groups? | `group_by()`|

Here is the dataset -- all flights originating and departing from Columbus (Ohio) January through September of 2017. 

Let  us load the data, and the `{tidyverse}` and `{here}` packages. 

In [1]:
library(tidyverse) 

load("data/cmhflights_01092017.RData")

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.2     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.2     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.1     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


In [2]:
names(cmhflights) # will show you the column names

[1] "Year"                 "Quarter"              "Month"               
  [4] "DayofMonth"           "DayOfWeek"            "FlightDate"          
  [7] "UniqueCarrier"        "AirlineID"            "Carrier"             
 [10] "TailNum"              "FlightNum"            "OriginAirportID"     
 [13] "OriginAirportSeqID"   "OriginCityMarketID"   "Origin"              
 [16] "OriginCityName"       "OriginState"          "OriginStateFips"     
 [19] "OriginStateName"      "OriginWac"            "DestAirportID"       
 [22] "DestAirportSeqID"     "DestCityMarketID"     "Dest"                
 [25] "DestCityName"         "DestState"            "DestStateFips"       
 [28] "DestStateName"        "DestWac"              "CRSDepTime"          
 [31] "DepTime"              "DepDelay"             "DepDelayMinutes"     
 [34] "DepDel15"             "DepartureDelayGroups" "DepTimeBlk"          
 [37] "TaxiOut"              "WheelsOff"            "WheelsOn"            
 [40] "TaxiIn"               "CRSArrTime"           "ArrTime"             
 [43] "ArrDelay"             "ArrDelayMinutes"      "ArrDel15"            
 [46] "ArrivalDelayGroups"   "ArrTimeBlk"           "Cancelled"           
 [49] "CancellationCode"     "Diverted"             "CRSElapsedTime"      
 [52] "ActualElapsedTime"    "AirTime"              "Flights"             
 [55] "Distance"             "DistanceGroup"        "CarrierDelay"        
 [58] "WeatherDelay"         "NASDelay"             "SecurityDelay"       
 [61] "LateAircraftDelay"    "FirstDepTime"         "TotalAddGTime"       
 [64] "LongestAddGTime"      "DivAirportLandings"   "DivReachedDest"      
 [67] "DivActualElapsedTime" "DivArrDelay"          "DivDistance"         
 [70] "Div1Airport"          "Div1AirportID"        "Div1AirportSeqID"    
 [73] "Div1WheelsOn"         "Div1TotalGTime"       "Div1LongestGTime"    
 [76] "Div1WheelsOff"        "Div1TailNum"          "Div2Airport"         
 [79] "Div2AirportID"        "Div2AirportSeqID"     "Div2WheelsOn"        
 [82] "Div2TotalGTime"       "Div2LongestGTime"     "Div2WheelsOff"       
 [85] "Div2TailNum"          "Div3Airport"          "Div3AirportID"       
 [88] "Div3AirportSeqID"     "Div3WheelsOn"         "Div3TotalGTime"      
 [91] "Div3LongestGTime"     "Div3WheelsOff"        "Div3TailNum"         
 [94] "Div4Airport"          "Div4AirportID"        "Div4AirportSeqID"    
 [97] "Div4WheelsOn"         "Div4TotalGTime"       "Div4LongestGTime"    
[100] "Div4WheelsOff"        "Div4TailNum"          "Div5Airport"         
[103] "Div5AirportID"        "Div5AirportSeqID"     "Div5WheelsOn"        
[106] "Div5TotalGTime"       "Div5LongestGTime"     "Div5WheelsOff"       
[109] "Div5TailNum"          "X110"

The output is rather onerous with 110 columns displayed, the last being `X110`, an empty column that can be dropped. 

In [3]:
cmhflights$X110 <- NULL #This command will delete the column named X110 

Did it work? Let us check.

In [4]:
names(cmhflights)

[1] "Year"                 "Quarter"              "Month"               
  [4] "DayofMonth"           "DayOfWeek"            "FlightDate"          
  [7] "UniqueCarrier"        "AirlineID"            "Carrier"             
 [10] "TailNum"              "FlightNum"            "OriginAirportID"     
 [13] "OriginAirportSeqID"   "OriginCityMarketID"   "Origin"              
 [16] "OriginCityName"       "OriginState"          "OriginStateFips"     
 [19] "OriginStateName"      "OriginWac"            "DestAirportID"       
 [22] "DestAirportSeqID"     "DestCityMarketID"     "Dest"                
 [25] "DestCityName"         "DestState"            "DestStateFips"       
 [28] "DestStateName"        "DestWac"              "CRSDepTime"          
 [31] "DepTime"              "DepDelay"             "DepDelayMinutes"     
 [34] "DepDel15"             "DepartureDelayGroups" "DepTimeBlk"          
 [37] "TaxiOut"              "WheelsOff"            "WheelsOn"            
 [40] "TaxiIn"               "CRSArrTime"           "ArrTime"             
 [43] "ArrDelay"             "ArrDelayMinutes"      "ArrDel15"            
 [46] "ArrivalDelayGroups"   "ArrTimeBlk"           "Cancelled"           
 [49] "CancellationCode"     "Diverted"             "CRSElapsedTime"      
 [52] "ActualElapsedTime"    "AirTime"              "Flights"             
 [55] "Distance"             "DistanceGroup"        "CarrierDelay"        
 [58] "WeatherDelay"         "NASDelay"             "SecurityDelay"       
 [61] "LateAircraftDelay"    "FirstDepTime"         "TotalAddGTime"       
 [64] "LongestAddGTime"      "DivAirportLandings"   "DivReachedDest"      
 [67] "DivActualElapsedTime" "DivArrDelay"          "DivDistance"         
 [70] "Div1Airport"          "Div1AirportID"        "Div1AirportSeqID"    
 [73] "Div1WheelsOn"         "Div1TotalGTime"       "Div1LongestGTime"    
 [76] "Div1WheelsOff"        "Div1TailNum"          "Div2Airport"         
 [79] "Div2AirportID"        "Div2AirportSeqID"     "Div2WheelsOn"        
 [82] "Div2TotalGTime"       "Div2LongestGTime"     "Div2WheelsOff"       
 [85] "Div2TailNum"          "Div3Airport"          "Div3AirportID"       
 [88] "Div3AirportSeqID"     "Div3WheelsOn"         "Div3TotalGTime"      
 [91] "Div3LongestGTime"     "Div3WheelsOff"        "Div3TailNum"         
 [94] "Div4Airport"          "Div4AirportID"        "Div4AirportSeqID"    
 [97] "Div4WheelsOn"         "Div4TotalGTime"       "Div4LongestGTime"    
[100] "Div4WheelsOff"        "Div4TailNum"          "Div5Airport"         
[103] "Div5AirportID"        "Div5AirportSeqID"     "Div5WheelsOn"        
[106] "Div5TotalGTime"       "Div5LongestGTime"     "Div5WheelsOff"       
[109] "Div5TailNum"

Now we are down to 109 columns and ready to get to work. 

## select() 
As in the present case of 100+ columns, often our data frame will have more columns than we plan to work with. In such instances it is a good idea to drop all unwanted columns; this will make the data-set more manageable and tax our computing resources less. For example, say I only want the first five columns (which happen to be `Year`, `Quarter`, `Month`, `DayofMonth`, and `DayOfWeek`). I could use `select` to create a data frame with only these columns:

In [5]:
cmhflights %>% 
  select(Year:DayOfWeek) -> my.df

The `:` is the bridge between __consecutive columns__ starting with `Year` and stopping with `DayofWeek`.

In addition, note that the subset of columns selected and all rows of data are being written to a new data frame called `my.df`

Quick check to see if we have only the columns we wanted. 

In [6]:
names(my.df)

[1] "Year"       "Quarter"    "Month"      "DayofMonth" "DayOfWeek"

Wonderful! It worked.

Now, the same result could have been obtained by taking the longer route of __listing each column by name__, as in the following: 

In [7]:
cmhflights %>%
  select(Year, Quarter, Month, DayofMonth, DayOfWeek) -> my.df 

names(my.df)

[1] "Year"       "Quarter"    "Month"      "DayofMonth" "DayOfWeek"

What if the columns were not sequentially located? In that case we would need to list each column we want. Say I want `Year`, `FlightDate`, `UniqueCarrier`, and `TailNum`. 

In [8]:
cmhflights %>% 
  select(Year, FlightDate:UniqueCarrier, TailNum) -> my.df 

names(my.df)

[1] "Year"          "FlightDate"    "UniqueCarrier" "TailNum"

Could we have used __column numbers instead of column names__? Absolutely. 

In [9]:
cmhflights %>% 
  select(c(1, 3, 5, 7)) -> my.df 

names(my.df)

[1] "Year"          "Month"         "DayOfWeek"     "UniqueCarrier"

You can also use __consecutive column numbers__, for examples, columns 1 through 5 as follows: 

In [10]:
cmhflights %>% 
  select(c(1:5)) -> my.df 

names(my.df)

[1] "Year"       "Quarter"    "Month"      "DayofMonth" "DayOfWeek"

You can also use column numbers to select a mix of columns, some that may be sequential and others that may be not sequential.

In [11]:
cmhflights %>% 
  select(c(1, 6:9, 12)) -> my.df

names(my.df)

[1] "Year"            "FlightDate"      "UniqueCarrier"   "AirlineID"      
[5] "Carrier"         "OriginAirportID"

### select() in other ways
We can also select columns in other ways, by specifying that  the column name `contain` some element. The code below shows you how this is done if I am looking for column names with the phrase "Carrier", then with "De", and then with "Num"  

In [12]:
cmhflights %>% 
  select(contains("Carrier")) -> my.df

names(my.df)

[1] "UniqueCarrier" "Carrier"       "CarrierDelay"

You can also specify that the columns to be selected __start with__ some alphanumeric string, for example, 

In [13]:
cmhflights %>% 
  select(starts_with("De")) -> my.df

names(my.df)

[1] "DestAirportID"        "DestAirportSeqID"     "DestCityMarketID"    
 [4] "Dest"                 "DestCityName"         "DestState"           
 [7] "DestStateFips"        "DestStateName"        "DestWac"             
[10] "DepTime"              "DepDelay"             "DepDelayMinutes"     
[13] "DepDel15"             "DepartureDelayGroups" "DepTimeBlk"

The other option would be to choose columns that __end with__ a particular alphanumeruc string, for example, 

In [14]:
cmhflights %>% 
  select(ends_with("Num")) -> my.df

names(my.df)

[1] "TailNum"     "FlightNum"   "Div1TailNum" "Div2TailNum" "Div3TailNum"
[6] "Div4TailNum" "Div5TailNum"

There are two other options -- `matches()` and `num_range()`. Let us look at each in turn. 

In [15]:
cmhflights %>%
    select(matches("el")) -> my.df

names(my.df)

[1] "DepDelay"             "DepDelayMinutes"      "DepDel15"            
 [4] "DepartureDelayGroups" "WheelsOff"            "WheelsOn"            
 [7] "ArrDelay"             "ArrDelayMinutes"      "ArrDel15"            
[10] "ArrivalDelayGroups"   "Cancelled"            "CancellationCode"    
[13] "CRSElapsedTime"       "ActualElapsedTime"    "CarrierDelay"        
[16] "WeatherDelay"         "NASDelay"             "SecurityDelay"       
[19] "LateAircraftDelay"    "DivActualElapsedTime" "DivArrDelay"         
[22] "Div1WheelsOn"         "Div1WheelsOff"        "Div2WheelsOn"        
[25] "Div2WheelsOff"        "Div3WheelsOn"         "Div3WheelsOff"       
[28] "Div4WheelsOn"         "Div4WheelsOff"        "Div5WheelsOn"        
[31] "Div5WheelsOff"

Here is another dataset where some of the column names contain a number.

In [16]:
head(billboard)

artist       track                   date.entered wk1 wk2 wk3 wk4 wk5 wk6 wk7
1 2 Pac        Baby Don't Cry (Keep... 2000-02-26   87  82  72  77  87  94  99 
2 2Ge+her      The Hardest Part Of ... 2000-09-02   91  87  92  NA  NA  NA  NA 
3 3 Doors Down Kryptonite              2000-04-08   81  70  68  67  66  57  54 
4 3 Doors Down Loser                   2000-10-21   76  76  72  69  67  65  55 
5 504 Boyz     Wobble Wobble           2000-04-15   57  34  25  17  17  31  36 
6 98^0         Give Me Just One Nig... 2000-08-19   51  39  34  26  26  19   2 
  ⋯ wk67 wk68 wk69 wk70 wk71 wk72 wk73 wk74 wk75 wk76
1 ⋯ NA   NA   NA   NA   NA   NA   NA   NA   NA   NA  
2 ⋯ NA   NA   NA   NA   NA   NA   NA   NA   NA   NA  
3 ⋯ NA   NA   NA   NA   NA   NA   NA   NA   NA   NA  
4 ⋯ NA   NA   NA   NA   NA   NA   NA   NA   NA   NA  
5 ⋯ NA   NA   NA   NA   NA   NA   NA   NA   NA   NA  
6 ⋯ NA   NA   NA   NA   NA   NA   NA   NA   NA   NA

In [17]:
billboard %>%
    select(num_range("wk", 1:5)) -> my.df

names(my.df)

[1] "wk1" "wk2" "wk3" "wk4" "wk5"

------------------

## filter()
Do you really want all the rows in the data-set or do you want to see only very specific rows that meet some criteria? Say we only want to look at certain months, or only flights on Saturdays and Sundays, or flights in a given month. For example, say we only want flights in January, i.e., `Month == 1`. 

In [18]:
cmhflights %>% 
  filter(Month == 1) -> my.df

table(my.df$Month) # Show me a frequency table for Month in my.df


   1 
3757 

What about only American Airline flights in January? Note that the UniqueCarrier code for this airline is AA.

In [19]:
cmhflights %>% 
  filter(Month == 1, UniqueCarrier == "AA") -> my.df 

table(my.df$Month, my.df$UniqueCarrier) # a simple frequency table

   
     AA
  1 387

Note that `,` inside `filter` means `&`

What about United Airlines flights in January to CMH (the airport code for Columbus, OH), regardless of where the flight originated? 

In [20]:
cmhflights %>% 
  filter(Month == 1, UniqueCarrier == "UA", Dest == "CMH") -> my.df

head(my.df)

Year Quarter Month DayofMonth DayOfWeek FlightDate UniqueCarrier AirlineID
1 2017 1       1     26         4         2017-01-26 UA            19977    
2 2017 1       1     26         4         2017-01-26 UA            19977    
3 2017 1       1     25         3         2017-01-25 UA            19977    
4 2017 1       1     25         3         2017-01-25 UA            19977    
5 2017 1       1     24         2         2017-01-24 UA            19977    
6 2017 1       1     24         2         2017-01-24 UA            19977    
  Carrier TailNum ⋯ Div4WheelsOff Div4TailNum Div5Airport Div5AirportID
1 UA      N889UA  ⋯ NA            NA          NA          NA           
2 UA      N39423  ⋯ NA            NA          NA          NA           
3 UA      N834UA  ⋯ NA            NA          NA          NA           
4 UA      N37462  ⋯ NA            NA          NA          NA           
5 UA      N855UA  ⋯ NA            NA          NA          NA           
6 UA      N75428  ⋯ NA            NA          NA          NA           
  Div5AirportSeqID Div5WheelsOn Div5TotalGTime Div5LongestGTime Div5WheelsOff
1 NA               NA           NA             NA               NA           
2 NA               NA           NA             NA               NA           
3 NA               NA           NA             NA               NA           
4 NA               NA           NA             NA               NA           
5 NA               NA           NA             NA               NA           
6 NA               NA           NA             NA               NA           
  Div5TailNum
1 NA         
2 NA         
3 NA         
4 NA         
5 NA         
6 NA

What if I wanted a more complicated filter, say, flights in January or February, either to CMH or to ORD (the airport code for O'Hare in Chicago)?

In [21]:
cmhflights %>% 
  filter(
    Month %in% c(1, 2), UniqueCarrier == "UA", Dest %in% c("CMH", "ORD")
    ) -> my.df 

table(my.df$Month) # frequency table of Month


  1   2 
106 145 

In [22]:
table(my.df$UniqueCarrier) # frequency table of UniqueCarrier


 UA 
251 

In [23]:
table(my.df$Dest) # frequency table of Dest


CMH ORD 
132 119 

In [24]:
cmhflights %>% 
  filter(
    !Month %in% c(1, 2)
    ) -> my.df 

table(my.df$Month)


   3    4    5    6    7    8    9 
4101 4123 4098 4138 4295 4279 3789 

In [25]:
cmhflights %>%
filter(
    Month != 1, Month != 2
    ) -> my.df

table(my.df$Month)


   3    4    5    6    7    8    9 
4101 4123 4098 4138 4295 4279 3789 

Beautiful, just beautiful. 

At this point it may not be readily apparent to you but using `%in% c(...)` makes applying complex filters easier than if you go some other route. 

Before we move on, note the operators  that work with `filter()` 

| **Operator** | **Meaning**  | **Operator** | **Meaning** |
| :---- | :---- | :---- |  :---- |
| $<$    | less than | $>$    | greater than |
| $==$   | equal to  | $\leq$ | less than or equal to |
| $\geq$ | greater than or equal to | `!=` | not equal to |
| `%in%`   | is a member of |  `is.na` | is NA |
| `!is.na` | is not NA  | `&,!,etc` | Boolean operators |

----------------

## arrange()

Now let us say I wanted to arrange the resulting data frame in `ascending order` of departure delays. How might I do that? Via `arrange()` 

Before we see this new command in action I am going to whittle the data-frame to only a few columns, and only flights to CMH or ORD. That will make it easier to see the result of executed commands.

In [26]:
cmhflights %>% 
    select('Year', 'Month', 'DayofMonth', 'FlightDate', 'Carrier', 'TailNum', 'FlightNum', 
           'Origin', 'OriginCityName', 'Dest', 'DestCityName', 'CRSDepTime', 'DepTime', 
           'DepDelay', 'DepDelayMinutes', 'CRSArrTime', 'ArrTime', 'ArrDelay', 'ArrDelayMinutes') %>%
    filter(Dest %in% c("CMH", "ORD")) -> my.df

my.df %>% 
  arrange(DepDelayMinutes) -> my.df2

my.df2

Year Month DayofMonth FlightDate Carrier TailNum FlightNum Origin
1     2017 1      1         2017-01-01 AA      N802AA  508       LAX   
2     2017 1      3         2017-01-03 AA      N760AA  508       LAX   
3     2017 1      4         2017-01-04 AA      N753AA  508       LAX   
4     2017 1      5         2017-01-05 AA      N744AA  508       LAX   
5     2017 1      7         2017-01-07 AA      N762AA  508       LAX   
6     2017 1      8         2017-01-08 AA      N808AA  508       LAX   
7     2017 1     11         2017-01-11 AA      N756AA  508       LAX   
8     2017 1     12         2017-01-12 AA      N805AA  508       LAX   
9     2017 1     13         2017-01-13 AA      N766AA  508       LAX   
10    2017 1     18         2017-01-18 AA      N825AA  508       LAX   
11    2017 1     25         2017-01-25 AA      N815AA  508       LAX   
12    2017 1     26         2017-01-26 AA      N755AA  508       LAX   
13    2017 1     27         2017-01-27 AA      N834AA  508       LAX   
14    2017 1     30         2017-01-30 AA      N815AA  508       LAX   
15    2017 1      1         2017-01-01 AA      N820AA  514       PHX   
16    2017 1      3         2017-01-03 AA      N742AA  514       PHX   
17    2017 1     11         2017-01-11 AA      N742AA  514       PHX   
18    2017 1     12         2017-01-12 AA      N762AA  514       PHX   
19    2017 1     13         2017-01-13 AA      N822AA  514       PHX   
20    2017 1     15         2017-01-15 AA      N817AA  514       PHX   
21    2017 1     16         2017-01-16 AA      N755AA  514       PHX   
22    2017 1     18         2017-01-18 AA      N840AA  514       PHX   
23    2017 1     19         2017-01-19 AA      N819AA  514       PHX   
24    2017 1     22         2017-01-22 AA      N833AA  514       PHX   
25    2017 1     23         2017-01-23 AA      N760AA  514       PHX   
26    2017 1     25         2017-01-25 AA      N823AA  514       PHX   
27    2017 1     26         2017-01-26 AA      N833AA  514       PHX   
28    2017 1     28         2017-01-28 AA      N755AA  514       PHX   
29    2017 1     29         2017-01-29 AA      N823AA  514       PHX   
30    2017 1     30         2017-01-30 AA      N760AA  514       PHX   
⋮     ⋮    ⋮     ⋮          ⋮          ⋮       ⋮       ⋮         ⋮     
18770 2017 9     20         2017-09-20 WN      N7726A   494      BWI   
18771 2017 9     10         2017-09-10 WN      NA      3675      MCO   
18772 2017 9     10         2017-09-10 WN      NA      4643      MCO   
18773 2017 9     10         2017-09-10 WN      N8327A  4930      MCO   
18774 2017 9     10         2017-09-10 WN      N430WN  3714      RSW   
18775 2017 9     10         2017-09-10 WN      N7818L  4454      TPA   
18776 2017 9     10         2017-09-10 WN      N915WN  4701      TPA   
18777 2017 9     11         2017-09-11 WN      N389SW    17      ATL   
18778 2017 9     11         2017-09-11 WN      N639SW   188      ATL   
18779 2017 9     11         2017-09-11 WN      N703SW  1816      DEN   
18780 2017 9     11         2017-09-11 WN      N7878A  1792      MCO   
18781 2017 9     11         2017-09-11 WN      N248WN  1888      MCO   
18782 2017 9     11         2017-09-11 WN      N553WN   213      MDW   
18783 2017 9     20         2017-09-20 WN      N947WN  5861      MDW   
18784 2017 9      5         2017-09-05 WN      N726SW   494      BWI   
18785 2017 9     11         2017-09-11 WN      NA       690      RSW   
18786 2017 9     11         2017-09-11 WN      N484WN   924      TPA   
18787 2017 9     12         2017-09-12 WN      NA       192      ATL   
18788 2017 9     12         2017-09-12 WN      NA      1623      DCA   
18789 2017 9     12         2017-09-12 WN      NA      1792      MCO   
18790 2017 9     12         2017-09-12 WN      NA       690      RSW   
18791 2017 9     12         2017-09-12 WN      N426WN   924      TPA   
18792 2017 9     13         2017-09-13 WN      NA       938      MDW   
18793 2017 9     13         2017-09-13 WN     

And now in `descending order` of delays by adding the minus symbol `-` to the  column name. 

In [27]:
my.df %>% 
  arrange(-DepDelayMinutes) -> my.df2

my.df2

Year Month DayofMonth FlightDate Carrier TailNum FlightNum Origin
1     2017 6     18         2017-06-18 EV      N18556  4080      CMH   
2     2017 6     29         2017-06-29 DL      N956DL  1897      ATL   
3     2017 8      9         2017-08-09 F9      N223FR  1686      LAS   
4     2017 1     29         2017-01-29 DL      N838DN  1897      ATL   
5     2017 4      8         2017-04-08 DL      N984DL  1276      ATL   
6     2017 1     23         2017-01-23 WN      N731SA  1886      DCA   
7     2017 2     19         2017-02-19 UA      N812UA   515      CMH   
8     2017 1     16         2017-01-16 EV      N607LR  5440      LGA   
9     2017 2     24         2017-02-24 UA      N849UA   688      ORD   
10    2017 8     17         2017-08-17 WN      N955WN  1031      STL   
11    2017 4     30         2017-04-30 WN      N937WN  5531      MSY   
12    2017 4      3         2017-04-03 DL      N930DL  1193      ATL   
13    2017 1     24         2017-01-24 EV      N691CA  5333      MSP   
14    2017 7     18         2017-07-18 WN      N7733B  3263      BOS   
15    2017 1     11         2017-01-11 WN      N492WN   254      MDW   
16    2017 3      7         2017-03-07 F9      N910FR  1694      MCO   
17    2017 6     14         2017-06-14 WN      N224WN  1639      ATL   
18    2017 4      3         2017-04-03 DL      N946DL  1152      ATL   
19    2017 5     24         2017-05-24 WN      N7707C   158      TPA   
20    2017 4      5         2017-04-05 WN      N635SW  2587      ATL   
21    2017 5     11         2017-05-11 AA      N4YLAA  2452      DFW   
22    2017 4      6         2017-04-06 WN      N429WN  1953      TPA   
23    2017 5      7         2017-05-07 WN      N8670A  2747      OAK   
24    2017 4      3         2017-04-03 DL      N329NW  1398      ATL   
25    2017 7      9         2017-07-09 AA      N812AA   508      LAX   
26    2017 7      2         2017-07-02 F9      N920FR  1694      MCO   
27    2017 3      9         2017-03-09 WN      N376SW  1782      BOS   
28    2017 9     11         2017-09-11 DL      N938DN  1152      ATL   
29    2017 4      9         2017-04-09 WN      N621SW  6325      BNA   
30    2017 6     15         2017-06-15 AA      N4XMAA   909      DFW   
⋮     ⋮    ⋮     ⋮          ⋮          ⋮       ⋮       ⋮         ⋮     
18770 2017 9     20         2017-09-20 WN      N7726A   494      BWI   
18771 2017 9     10         2017-09-10 WN      NA      3675      MCO   
18772 2017 9     10         2017-09-10 WN      NA      4643      MCO   
18773 2017 9     10         2017-09-10 WN      N8327A  4930      MCO   
18774 2017 9     10         2017-09-10 WN      N430WN  3714      RSW   
18775 2017 9     10         2017-09-10 WN      N7818L  4454      TPA   
18776 2017 9     10         2017-09-10 WN      N915WN  4701      TPA   
18777 2017 9     11         2017-09-11 WN      N389SW    17      ATL   
18778 2017 9     11         2017-09-11 WN      N639SW   188      ATL   
18779 2017 9     11         2017-09-11 WN      N703SW  1816      DEN   
18780 2017 9     11         2017-09-11 WN      N7878A  1792      MCO   
18781 2017 9     11         2017-09-11 WN      N248WN  1888      MCO   
18782 2017 9     11         2017-09-11 WN      N553WN   213      MDW   
18783 2017 9     20         2017-09-20 WN      N947WN  5861      MDW   
18784 2017 9      5         2017-09-05 WN      N726SW   494      BWI   
18785 2017 9     11         2017-09-11 WN      NA       690      RSW   
18786 2017 9     11         2017-09-11 WN      N484WN   924      TPA   
18787 2017 9     12         2017-09-12 WN      NA       192      ATL   
18788 2017 9     12         2017-09-12 WN      NA      1623      DCA   
18789 2017 9     12         2017-09-12 WN      NA      1792      MCO   
18790 2017 9     12         2017-09-12 WN      NA       690      RSW   
18791 2017 9     12         2017-09-12 WN      N426WN   924      TPA   
18792 2017 9     13         2017-09-13 WN      NA       938      MDW   
18793 2017 9     13         2017-09-13 WN     

We could tweak this further, perhaps saying sort by departure delays to CMH, and then to ORD. 

In [28]:
my.df %>% 
  arrange(Dest, -DepDelayMinutes) -> my.df2

my.df2

Year Month DayofMonth FlightDate Carrier TailNum FlightNum Origin
1     2017 6     29         2017-06-29 DL      N956DL  1897      ATL   
2     2017 8      9         2017-08-09 F9      N223FR  1686      LAS   
3     2017 1     29         2017-01-29 DL      N838DN  1897      ATL   
4     2017 4      8         2017-04-08 DL      N984DL  1276      ATL   
5     2017 1     23         2017-01-23 WN      N731SA  1886      DCA   
6     2017 1     16         2017-01-16 EV      N607LR  5440      LGA   
7     2017 2     24         2017-02-24 UA      N849UA   688      ORD   
8     2017 8     17         2017-08-17 WN      N955WN  1031      STL   
9     2017 4     30         2017-04-30 WN      N937WN  5531      MSY   
10    2017 4      3         2017-04-03 DL      N930DL  1193      ATL   
11    2017 1     24         2017-01-24 EV      N691CA  5333      MSP   
12    2017 7     18         2017-07-18 WN      N7733B  3263      BOS   
13    2017 1     11         2017-01-11 WN      N492WN   254      MDW   
14    2017 3      7         2017-03-07 F9      N910FR  1694      MCO   
15    2017 6     14         2017-06-14 WN      N224WN  1639      ATL   
16    2017 4      3         2017-04-03 DL      N946DL  1152      ATL   
17    2017 5     24         2017-05-24 WN      N7707C   158      TPA   
18    2017 4      5         2017-04-05 WN      N635SW  2587      ATL   
19    2017 5     11         2017-05-11 AA      N4YLAA  2452      DFW   
20    2017 4      6         2017-04-06 WN      N429WN  1953      TPA   
21    2017 5      7         2017-05-07 WN      N8670A  2747      OAK   
22    2017 4      3         2017-04-03 DL      N329NW  1398      ATL   
23    2017 7      9         2017-07-09 AA      N812AA   508      LAX   
24    2017 7      2         2017-07-02 F9      N920FR  1694      MCO   
25    2017 3      9         2017-03-09 WN      N376SW  1782      BOS   
26    2017 9     11         2017-09-11 DL      N938DN  1152      ATL   
27    2017 4      9         2017-04-09 WN      N621SW  6325      BNA   
28    2017 6     15         2017-06-15 AA      N4XMAA   909      DFW   
29    2017 6     19         2017-06-19 OO      N935SW  4520      DTW   
30    2017 7     14         2017-07-14 WN      N467WN  1922      DAL   
⋮     ⋮    ⋮     ⋮          ⋮          ⋮       ⋮       ⋮         ⋮     
18770 2017 9     21         2017-09-21 UA      N78506  1014      CMH   
18771 2017 9     20         2017-09-20 UA      N24729   515      CMH   
18772 2017 9     20         2017-09-20 UA      N35260  1014      CMH   
18773 2017 9     19         2017-09-19 UA      N16713   515      CMH   
18774 2017 9     19         2017-09-19 UA      N76533  1014      CMH   
18775 2017 9     18         2017-09-18 UA      N24729   515      CMH   
18776 2017 9      3         2017-09-03 UA      N892UA   515      CMH   
18777 2017 9      2         2017-09-02 UA      N463UA   515      CMH   
18778 2017 9      2         2017-09-02 UA      N842UA   805      CMH   
18779 2017 9      1         2017-09-01 UA      N38268   515      CMH   
18780 2017 9     18         2017-09-18 UA      N73276  1014      CMH   
18781 2017 9     15         2017-09-15 UA      N24729   515      CMH   
18782 2017 9     15         2017-09-15 UA      N13248  1014      CMH   
18783 2017 9     14         2017-09-14 UA      N17752   515      CMH   
18784 2017 9      1         2017-09-01 UA      N17753  1802      CMH   
18785 2017 9      9         2017-09-09 EV      N13550  3795      CMH   
18786 2017 9     16         2017-09-16 EV      N13913  3795      CMH   
18787 2017 9     23         2017-09-23 EV      N14993  3795      CMH   
18788 2017 9     30         2017-09-30 EV      N26545  3795      CMH   
18789 2017 5     18         2017-05-18 UA      NA       245      CMH   
18790 2017 6     13         2017-06-13 OO      N141SY  5551      CMH   
18791 2017 6     14         2017-06-14 OO      N128SY  5551      CMH   
18792 2017 7     20         2017-07-20 OO      N120SY  5386      CMH   
18793 2017 7     21         2017-07-21 OO     

So far, we have seen each function in isolation. Now we streamline things a bit so that we only end up with the columns and rows we want to work with, arranged as we want the resulting data-set to be. 

In [29]:
cmhflights %>% 
  select(Month, UniqueCarrier, Dest, DepDelayMinutes) %>% 
  filter(
    Month %in% c(1, 2), UniqueCarrier == "UA", Dest %in% c("CMH", "ORD")
    ) %>% 
  arrange(Month, Dest, -DepDelayMinutes) -> my.df3

my.df3

Month UniqueCarrier Dest DepDelayMinutes
1   1     UA            CMH  178            
2   1     UA            CMH   61            
3   1     UA            CMH   44            
4   1     UA            CMH   39            
5   1     UA            CMH   39            
6   1     UA            CMH   30            
7   1     UA            CMH   27            
8   1     UA            CMH   18            
9   1     UA            CMH   12            
10  1     UA            CMH   10            
11  1     UA            CMH    8            
12  1     UA            CMH    2            
13  1     UA            CMH    2            
14  1     UA            CMH    2            
15  1     UA            CMH    1            
16  1     UA            CMH    1            
17  1     UA            CMH    1            
18  1     UA            CMH    0            
19  1     UA            CMH    0            
20  1     UA            CMH    0            
21  1     UA            CMH    0            
22  1     UA            CMH    0            
23  1     UA            CMH    0            
24  1     UA            CMH    0            
25  1     UA            CMH    0            
26  1     UA            CMH    0            
27  1     UA            CMH    0            
28  1     UA            CMH    0            
29  1     UA            CMH    0            
30  1     UA            CMH    0            
⋮   ⋮     ⋮             ⋮    ⋮              
222 2     UA            ORD  0              
223 2     UA            ORD  0              
224 2     UA            ORD  0              
225 2     UA            ORD  0              
226 2     UA            ORD  0              
227 2     UA            ORD  0              
228 2     UA            ORD  0              
229 2     UA            ORD  0              
230 2     UA            ORD  0              
231 2     UA            ORD  0              
232 2     UA            ORD  0              
233 2     UA            ORD  0              
234 2     UA            ORD  0              
235 2     UA            ORD  0              
236 2     UA            ORD  0              
237 2     UA            ORD  0              
238 2     UA            ORD  0              
239 2     UA            ORD  0              
240 2     UA            ORD  0              
241 2     UA            ORD  0              
242 2     UA            ORD  0              
243 2     UA            ORD  0              
244 2     UA            ORD  0              
245 2     UA            ORD  0              
246 2     UA            ORD  0              
247 2     UA            ORD  0              
248 2     UA            ORD  0              
249 2     UA            ORD  0              
250 2     UA            ORD  0              
251 2     UA            ORD  0

Here, the end result is a data frame arranged by Month, then within Month by Destination, and then finally by descending order of flight delays. This is the beauty of `dplyr`, allowing us to chain together various functions to get what we want. How is this helpful? Well, now you have a data frame that you can analyze. What do I want to calculate? Well, let us say we want to create a frequency table, something that we would have done with Excel via a Pivot Table. 

------------

## summarise() or summarize()
What if we need to calculate frequencies? For example, how many flights per month are there? What if we want the __mean__ DepDelay or __median__ ArrDelay? These can be easily calculated as shown below. 

Let us start with frequencies.

In [30]:
cmhflights %>%
  count(Month) # Most flights were in July (n = 4295)

Month n   
1 1     3757
2 2     3413
3 3     4101
4 4     4123
5 5     4098
6 6     4138
7 7     4295
8 8     4279
9 9     3789

What about by days of the week AND by month? 

In [31]:
cmhflights %>%
  count(Month, DayOfWeek) # Output is sorted by Month and then DayOfWeek

Month DayOfWeek n  
1  1     1         660
2  1     2         635
3  1     3         516
4  1     4         522
5  1     5         523
6  1     6         361
7  1     7         540
8  2     1         527
9  2     2         506
10 2     3         516
11 2     4         531
12 2     5         528
13 2     6         371
14 2     7         434
15 3     1         549
16 3     2         546
17 3     3         674
18 3     4         695
19 3     5         685
20 3     6         447
21 3     7         505
22 4     1         584
23 4     2         569
24 4     3         570
25 4     4         580
26 4     5         576
27 4     6         581
28 4     7         663
29 5     1         683
30 5     2         685
⋮  ⋮     ⋮         ⋮  
34 5     6         418
35 5     7         507
36 6     1         561
37 6     2         567
38 6     3         572
39 6     4         701
40 6     5         701
41 6     6         482
42 6     7         554
43 7     1         703
44 7     2         549
45 7     3         584
46 7     4         578
47 7     5         574
48 7     6         633
49 7     7         674
50 8     1         580
51 8     2         696
52 8     3         712
53 8     4         718
54 8     5         581
55 8     6         458
56 8     7         534
57 9     1         548
58 9     2         531
59 9     3         539
60 9     4         550
61 9     5         689
62 9     6         451
63 9     7         481

I want to know the average departure delay and the average arrival delay for all flights, with the averages calculated in two ways -- as the `mean` and as the `median`. Maybe I also want the variance and the standard deviation of both delays. 

In [32]:
cmhflights %>%
  summarise(
      mean_arr_delay = mean(ArrDelay, na.rm = TRUE),
      mean_dep_delay = mean(DepDelay, na.rm = TRUE),
      median_arr_delay = median(ArrDelay, na.rm = TRUE),
      median_dep_delay = median(DepDelay, na.rm = TRUE),
      variance_arr_delay = var(ArrDelay, na.rm = TRUE),
      variance_dep_delay = var(DepDelay, na.rm = TRUE),  
      sd_arr_delay = sd(ArrDelay, na.rm = TRUE),
      sd_dep_delay = sd(DepDelay, na.rm = TRUE)
  )

mean_arr_delay mean_dep_delay median_arr_delay median_dep_delay
1 3.258327       9.202072       -6               -2              
  variance_arr_delay variance_dep_delay sd_arr_delay sd_dep_delay
1 1746.284           1581.747           41.78856     39.77119

Here, the `na.rm = TRUE` command is useful because R will not allow you to calculate any mean, median, variance, etc. if the clumn includes some rows with missing data. 

You can see this below, where I have a small data-set called `df` with 4 values of x, but one of the four is missing and recorded as `NA`. See what happens when I try to calculate the mean with/without `na.rm = TRUE`.

In [33]:
df = data.frame(x = c(2, 4, 9, NA))

df

x 
1  2
2  4
3  9
4 NA

In [34]:
df %>%
  summarise(mean.x = mean(x)) # You get no meaningful values 

mean.x
1 NA

In [35]:
df %>%
  summarise(mean.x = mean(x, na.rm = TRUE)) # Now you do get something meaningful

mean.x
1 5

What if I want to count total DepDelay by airline? 

In [36]:
cmhflights %>%
    group_by(Carrier) %>%
    summarize(sum(DepDelay, na.rm = TRUE))

Carrier sum(DepDelay, na.rm = TRUE)
1 AA       29853                     
2 DL       38297                     
3 EV       44146                     
4 F9       16834                     
5 OO       13006                     
6 UA        9739                     
7 WN      175001

--------------

## group_by()
These summaries are fine if you want to calculate aggregate quantities of interest (i.e., means, medians, frequencies, variances, etc. for all rows of data) but what if you wanted to calculate the __number of flights per month, by airline? Average delays by airline?__ 

Now things get interesting because `group_by()` will open up this new world for us! The first thing I will calculate is the number of flights by airline per month. 

In [37]:
cmhflights %>%
  group_by(Month, Carrier) %>%
  tally()

Month Carrier n   
1  1     AA       387
2  1     DL       549
3  1     EV       436
4  1     F9       152
5  1     OO       123
6  1     UA       106
7  1     WN      2004
8  2     AA       360
9  2     DL       511
10 2     EV       362
11 2     F9       136
12 2     OO       100
13 2     UA       156
14 2     WN      1788
15 3     AA       418
16 3     DL       543
17 3     EV       408
18 3     F9       170
19 3     OO       116
20 3     UA       193
21 3     WN      2253
22 4     AA       442
23 4     DL       554
24 4     EV       316
25 4     F9       192
26 4     OO       288
27 4     UA       142
28 4     WN      2189
29 5     AA       442
30 5     DL       694
⋮  ⋮     ⋮       ⋮   
34 5     UA       178
35 5     WN      2052
36 6     AA       485
37 6     DL       646
38 6     EV       359
39 6     F9       180
40 6     OO       180
41 6     UA       184
42 6     WN      2104
43 7     AA       500
44 7     DL       674
45 7     EV       251
46 7     F9       186
47 7     OO       316
48 7     UA       194
49 7     WN      2174
50 8     AA       488
51 8     DL       666
52 8     EV       348
53 8     F9       186
54 8     OO       336
55 8     UA       178
56 8     WN      2077
57 9     AA       369
58 9     DL       609
59 9     EV       294
60 9     F9       180
61 9     OO       318
62 9     UA       196
63 9     WN      1823

You can get the same result if you use `summarize()`

In [38]:
cmhflights %>%
  group_by(Month, Carrier) %>%
  summarize(
    frequency = n()
    )

`summarise()` has grouped output by 'Month'. You can override using the
`.groups` argument.


Month Carrier frequency
1  1     AA       387     
2  1     DL       549     
3  1     EV       436     
4  1     F9       152     
5  1     OO       123     
6  1     UA       106     
7  1     WN      2004     
8  2     AA       360     
9  2     DL       511     
10 2     EV       362     
11 2     F9       136     
12 2     OO       100     
13 2     UA       156     
14 2     WN      1788     
15 3     AA       418     
16 3     DL       543     
17 3     EV       408     
18 3     F9       170     
19 3     OO       116     
20 3     UA       193     
21 3     WN      2253     
22 4     AA       442     
23 4     DL       554     
24 4     EV       316     
25 4     F9       192     
26 4     OO       288     
27 4     UA       142     
28 4     WN      2189     
29 5     AA       442     
30 5     DL       694     
⋮  ⋮     ⋮       ⋮        
34 5     UA       178     
35 5     WN      2052     
36 6     AA       485     
37 6     DL       646     
38 6     EV       359     
39 6     F9       180     
40 6     OO       180     
41 6     UA       184     
42 6     WN      2104     
43 7     AA       500     
44 7     DL       674     
45 7     EV       251     
46 7     F9       186     
47 7     OO       316     
48 7     UA       194     
49 7     WN      2174     
50 8     AA       488     
51 8     DL       666     
52 8     EV       348     
53 8     F9       186     
54 8     OO       336     
55 8     UA       178     
56 8     WN      2077     
57 9     AA       369     
58 9     DL       609     
59 9     EV       294     
60 9     F9       180     
61 9     OO       318     
62 9     UA       196     
63 9     WN      1823

Top recap, two ways to do this, first with `tally()` and the second with `summarize(frequency = n())` ... both yielding the same result. Remember `tally()` because it is shorter code. 

Now I want a table that gives us the number of flights per month per airline per destination. 

In [39]:
cmhflights %>%
  group_by(Month, Carrier, Dest) %>%
  tally()

Month Carrier Dest n  
1   1     AA      CMH  193
2   1     AA      DFW  112
3   1     AA      LAX   24
4   1     AA      PHX   58
5   1     DL      ATL  224
6   1     DL      CMH  275
7   1     DL      DTW    2
8   1     DL      LAX   27
9   1     DL      MSP   21
10  1     EV      CMH  218
11  1     EV      EWR   82
12  1     EV      IAH   72
13  1     EV      LGA   42
14  1     EV      MSP   18
15  1     EV      ORD    4
16  1     F9      CMH   76
17  1     F9      DEN   31
18  1     F9      LAS   31
19  1     F9      MCO   14
20  1     OO      CMH   61
21  1     OO      DTW   56
22  1     OO      IAH    1
23  1     OO      MSP    2
24  1     OO      ORD    3
25  1     UA      CMH   53
26  1     UA      ORD   53
27  1     WN      ATL   84
28  1     WN      BNA   54
29  1     WN      BOS   49
30  1     WN      BWI   90
⋮   ⋮     ⋮       ⋮    ⋮  
369 9     F9      DEN   30
370 9     F9      LAS   30
371 9     F9      MCO   30
372 9     OO      CMH  159
373 9     OO      DEN    3
374 9     OO      DTW   65
375 9     OO      MSP   21
376 9     OO      ORD   70
377 9     UA      CMH   98
378 9     UA      DEN   49
379 9     UA      IAH    1
380 9     UA      ORD   48
381 9     WN      ATL   80
382 9     WN      BNA   55
383 9     WN      BOS   54
384 9     WN      BWI  109
385 9     WN      CMH  912
386 9     WN      DAL   30
387 9     WN      DCA   55
388 9     WN      DEN   32
389 9     WN      FLL    5
390 9     WN      LAS   58
391 9     WN      MCO   65
392 9     WN      MDW  187
393 9     WN      MSY    4
394 9     WN      OAK   30
395 9     WN      PHX   30
396 9     WN      RSW   30
397 9     WN      STL   54
398 9     WN      TPA   33

We could keep enriching the grouping structure. For example, let us add the day of the week to the mix ... 

In [40]:
cmhflights %>%
  group_by(Month, Carrier, Dest, DayOfWeek) %>%
  tally()

Month Carrier Dest DayOfWeek n 
1    1     AA      CMH  1         35
2    1     AA      CMH  2         23
3    1     AA      CMH  3         28
4    1     AA      CMH  4         28
5    1     AA      CMH  5         28
6    1     AA      CMH  6         21
7    1     AA      CMH  7         30
8    1     AA      DFW  1         20
9    1     AA      DFW  2         16
10   1     AA      DFW  3         16
11   1     AA      DFW  4         16
12   1     AA      DFW  5         16
13   1     AA      DFW  6          9
14   1     AA      DFW  7         19
15   1     AA      LAX  1          5
16   1     AA      LAX  2          1
17   1     AA      LAX  3          4
18   1     AA      LAX  4          4
19   1     AA      LAX  5          4
20   1     AA      LAX  6          1
21   1     AA      LAX  7          5
22   1     AA      PHX  1         10
23   1     AA      PHX  2          6
24   1     AA      PHX  3          8
25   1     AA      PHX  4          8
26   1     AA      PHX  5          8
27   1     AA      PHX  6          8
28   1     AA      PHX  7         10
29   1     DL      ATL  1         40
30   1     DL      ATL  2         40
⋮    ⋮     ⋮       ⋮    ⋮         ⋮ 
2507 9     WN      OAK  6          5
2508 9     WN      OAK  7          4
2509 9     WN      PHX  1          4
2510 9     WN      PHX  2          4
2511 9     WN      PHX  3          4
2512 9     WN      PHX  4          4
2513 9     WN      PHX  5          5
2514 9     WN      PHX  6          5
2515 9     WN      PHX  7          4
2516 9     WN      RSW  1          4
2517 9     WN      RSW  2          4
2518 9     WN      RSW  3          4
2519 9     WN      RSW  4          4
2520 9     WN      RSW  5          5
2521 9     WN      RSW  6          5
2522 9     WN      RSW  7          4
2523 9     WN      STL  1          8
2524 9     WN      STL  2          8
2525 9     WN      STL  3          8
2526 9     WN      STL  4          8
2527 9     WN      STL  5         10
2528 9     WN      STL  6          5
2529 9     WN      STL  7          7
2530 9     WN      TPA  1          4
2531 9     WN      TPA  2          4
2532 9     WN      TPA  3          4
2533 9     WN      TPA  4          4
2534 9     WN      TPA  5          5
2535 9     WN      TPA  6          5
2536 9     WN      TPA  7          7

Now say I am really curious about mean departure delays for the preceding grouping structure. That is, what does mean departure delay look like for flights by day of the week, by month, by carrier, and by destination?

In [41]:
cmhflights %>%
  group_by(Month, Carrier, Dest, DayOfWeek) %>%
  summarise(mean_dep_delay = mean(DepDelay, na.rm = TRUE))

`summarise()` has grouped output by 'Month', 'Carrier', 'Dest'. You can
override using the `.groups` argument.


Month Carrier Dest DayOfWeek mean_dep_delay
1    1     AA      CMH  1         16.2857143    
2    1     AA      CMH  2          0.3043478    
3    1     AA      CMH  3          1.1428571    
4    1     AA      CMH  4          3.1071429    
5    1     AA      CMH  5          4.1071429    
6    1     AA      CMH  6         24.8571429    
7    1     AA      CMH  7         21.8333333    
8    1     AA      DFW  1         25.4000000    
9    1     AA      DFW  2         -2.6250000    
10   1     AA      DFW  3         17.6000000    
11   1     AA      DFW  4         -0.1875000    
12   1     AA      DFW  5          3.8750000    
13   1     AA      DFW  6         35.6666667    
14   1     AA      DFW  7         50.2105263    
15   1     AA      LAX  1          3.4000000    
16   1     AA      LAX  2         -4.0000000    
17   1     AA      LAX  3         -0.5000000    
18   1     AA      LAX  4          2.5000000    
19   1     AA      LAX  5         -1.5000000    
20   1     AA      LAX  6         -3.0000000    
21   1     AA      LAX  7         -5.8000000    
22   1     AA      PHX  1          1.7000000    
23   1     AA      PHX  2         -4.6666667    
24   1     AA      PHX  3         -0.1250000    
25   1     AA      PHX  4         -0.7500000    
26   1     AA      PHX  5         -0.8750000    
27   1     AA      PHX  6         -5.6250000    
28   1     AA      PHX  7         27.9000000    
29   1     DL      ATL  1         29.9459459    
30   1     DL      ATL  2          1.1500000    
⋮    ⋮     ⋮       ⋮    ⋮         ⋮             
2507 9     WN      OAK  6          4.4000000    
2508 9     WN      OAK  7          1.7500000    
2509 9     WN      PHX  1         22.0000000    
2510 9     WN      PHX  2         13.5000000    
2511 9     WN      PHX  3         29.5000000    
2512 9     WN      PHX  4         16.7500000    
2513 9     WN      PHX  5          8.0000000    
2514 9     WN      PHX  6         -1.4000000    
2515 9     WN      PHX  7         15.7500000    
2516 9     WN      RSW  1         -0.3333333    
2517 9     WN      RSW  2          2.3333333    
2518 9     WN      RSW  3         -5.3333333    
2519 9     WN      RSW  4         16.7500000    
2520 9     WN      RSW  5         22.6000000    
2521 9     WN      RSW  6         -1.2500000    
2522 9     WN      RSW  7         -1.6666667    
2523 9     WN      STL  1         20.2500000    
2524 9     WN      STL  2         -3.0000000    
2525 9     WN      STL  3         -3.8750000    
2526 9     WN      STL  4         14.3750000    
2527 9     WN      STL  5         -1.2000000    
2528 9     WN      STL  6         -3.2000000    
2529 9     WN      STL  7         -5.5714286    
2530 9     WN      TPA  1          5.3333333    
2531 9     WN      TPA  2          7.0000000    
2532 9     WN      TPA  3          2.0000000    
2533 9     WN      TPA  4         13.0000000    
2534 9     WN      TPA  5         14.8000000    
2535 9     WN      TPA  6          0.0000000    
2536 9     WN      TPA  7          6.8000000

But this is a complicated summary table. What if all I really want to know is what airline has the __highest mean departure delays__, regardless of month or destination or day of the week? This could be done by using `arrange()` to display the result in descending order of _mean_dep_delay_

In [42]:
cmhflights %>%
  group_by(Carrier) %>%
  summarise(mean_dep_delay = mean(DepDelay, na.rm = TRUE)) %>%
  arrange(-mean_dep_delay) # ordered in descending order of delays

Carrier mean_dep_delay
1 EV      15.207027     
2 F9      10.839665     
3 WN       9.581221     
4 AA       7.733938     
5 DL       7.095979     
6 OO       6.438614     
7 UA       6.390420

EV is Express Jet; F9 is Frontier Airlines; WN is Southwest Airlines; OO is SkyWest Airlines; AA is American Airlines; DL is Delta Airlines; UA is United Airlines. So clearly United Airlines had the lowest average departure delays. Would this still be true if we repeated the calculation by Month? 

In [43]:
cmhflights %>%
  group_by(Carrier, Month) %>%
  summarise(mean_dep_delay = mean(DepDelay, na.rm = TRUE)) %>%
  arrange(mean_dep_delay) # ordered in descending order of delays

`summarise()` has grouped output by 'Carrier'. You can override using the
`.groups` argument.


Carrier Month mean_dep_delay
1  OO      8     -0.8727273    
2  OO      2     -0.5700000    
3  OO      9     -0.1792453    
4  UA      9      0.6989796    
5  DL      8      1.8528529    
6  DL      2      2.2367906    
7  AA      2      2.3027778    
8  F9      9      2.5086705    
9  AA      9      2.6250000    
10 DL      3      2.9815838    
11 UA      4      3.1690141    
12 UA      8      3.5337079    
13 WN      2      3.6589758    
14 F9      2      3.8308824    
15 DL      9      3.9107438    
16 DL      5      3.9927954    
17 UA      5      4.2881356    
18 UA      1      4.5754717    
19 UA      3      4.8704663    
20 F9      4      5.6789474    
21 WN      3      6.0947226    
22 UA      2      6.2243590    
23 OO      4      6.4895833    
24 AA      5      6.5609195    
25 WN      9      6.6165627    
26 AA      8      6.8822314    
27 F9      5      7.4130435    
28 AA      4      7.7597254    
29 AA      3      7.8851675    
30 WN      1      8.1796165    
⋮  ⋮       ⋮     ⋮             
34 WN      5     10.30390      
35 OO      5     10.43561      
36 EV      8     10.80317      
37 OO      6     10.93750      
38 AA      1     10.99482      
39 AA      6     11.17537      
40 OO      3     11.21053      
41 DL      7     11.28338      
42 AA      7     11.33671      
43 F9      6     11.98333      
44 WN      8     12.13687      
45 EV      2     12.31858      
46 OO      7     12.73290      
47 DL      1     13.04833      
48 OO      1     13.59350      
49 EV      5     14.05495      
50 UA      6     14.09239      
51 F9      8     14.10870      
52 WN      6     14.18984      
53 WN      7     14.19621      
54 UA      7     14.45312      
55 F9      3     14.67059      
56 F9      7     14.96774      
57 EV      6     15.17094      
58 EV      3     16.45845      
59 EV      1     16.86744      
60 DL      4     17.72500      
61 EV      4     17.77199      
62 F9      1     22.70000      
63 EV      7     25.47034

All righty then! Looks like three of the lowest mean departure delays were for SkyWest. Do not let the negative numbers throw you for a loop; a negative value implies the flight left earlier than scheduled. 

So far so good. But now I am curious about what percent of flights operated by AA, DL, UA, and WN were delayed. How could I calculate this? 

(1) I need to use `filter()` to restrict the data-set to just these four airlines.  
(2) Then I need to generate a new column that identifies whether a flight was delayed or not (`late`).  
(3) Now we can calculate the total number of flights (`nflights`) and the total number of flights that were delayed (`nlate`).  
(4) If I then calculate $\left( \dfrac{nlate}{nflights} \right)\times 100$ we will end up with the percent of flights that were delayed.  

In [44]:
cmhflights %>%
  select(c(Carrier, DepDelay)) %>%
  filter(
      Carrier %in% c("AA", "DL", "UA", "WN")
  ) %>%
  mutate(
      late = case_when(
          DepDelay > 0 ~ "Yes",
          DepDelay <= 0 ~ "No"
      )
  ) %>% 
  group_by(Carrier) %>%
  mutate(
      nflights = n()
  ) %>%
  group_by(Carrier, late) %>%
  mutate(
    nlate = n(),
    pct_late = (nlate / nflights) * 100
  ) -> df1

df1

Carrier DepDelay late nflights nlate pct_late
1     AA       -9      No   3891     2698  69.33950
2     AA       24      Yes  3891     1162  29.86379
3     AA       -6      No   3891     2698  69.33950
4     AA       -5      No   3891     2698  69.33950
5     AA       -7      No   3891     2698  69.33950
6     AA       22      Yes  3891     1162  29.86379
7     AA       -4      No   3891     2698  69.33950
8     AA       -4      No   3891     2698  69.33950
9     AA       42      Yes  3891     1162  29.86379
10    AA       -6      No   3891     2698  69.33950
11    AA       -5      No   3891     2698  69.33950
12    AA       -3      No   3891     2698  69.33950
13    AA      106      Yes  3891     1162  29.86379
14    AA        4      Yes  3891     1162  29.86379
15    AA        0      No   3891     2698  69.33950
16    AA        6      Yes  3891     1162  29.86379
17    AA        7      Yes  3891     1162  29.86379
18    AA       22      Yes  3891     1162  29.86379
19    AA       31      Yes  3891     1162  29.86379
20    AA       -8      No   3891     2698  69.33950
21    AA       -5      No   3891     2698  69.33950
22    AA       -3      No   3891     2698  69.33950
23    AA       32      Yes  3891     1162  29.86379
24    AA       -3      No   3891     2698  69.33950
25    AA       -3      No   3891     2698  69.33950
26    AA       61      Yes  3891     1162  29.86379
27    AA       -4      No   3891     2698  69.33950
28    AA        8      Yes  3891     1162  29.86379
29    AA        0      No   3891     2698  69.33950
30    AA       -3      No   3891     2698  69.33950
⋮     ⋮       ⋮        ⋮    ⋮        ⋮     ⋮       
29299 DL      -2       No   5446     3887  71.37349
29300 DL      87       Yes  5446     1510  27.72677
29301 DL      -6       No   5446     3887  71.37349
29302 DL      36       Yes  5446     1510  27.72677
29303 DL      -3       No   5446     3887  71.37349
29304 DL      -5       No   5446     3887  71.37349
29305 DL       6       Yes  5446     1510  27.72677
29306 DL      -7       No   5446     3887  71.37349
29307 DL      -6       No   5446     3887  71.37349
29308 DL      -8       No   5446     3887  71.37349
29309 DL      -2       No   5446     3887  71.37349
29310 DL      -3       No   5446     3887  71.37349
29311 DL      -3       No   5446     3887  71.37349
29312 DL      -6       No   5446     3887  71.37349
29313 DL      -4       No   5446     3887  71.37349
29314 DL      -3       No   5446     3887  71.37349
29315 DL      -3       No   5446     3887  71.37349
29316 DL      -3       No   5446     3887  71.37349
29317 DL      -4       No   5446     3887  71.37349
29318 DL       9       Yes  5446     1510  27.72677
29319 DL      -1       No   5446     3887  71.37349
29320 DL      -5       No   5446     3887  71.37349
29321 DL      -3       No   5446     3887  71.37349
29322 DL      -2       No   5446     3887  71.37349
29323 DL      -3       No   5446     3887  71.37349
29324 DL      -2       No   5446     3887  71.37349
29325 DL      -6       No   5446     3887  71.37349
29326 DL       1       Yes  5446     1510  27.72677
29327 DL      -5       No   5446     3887  71.37349
29328 DL      -4       No   5446     3887  71.37349

There is a whole lot going on here so let us break it down. 

`filter(Carrier %in% c("AA", "DL", "UA", "WN"))` is keeping specified airlines' data while dropping the rest

`mutate(late = case_when(...)` is creating a new column called `late` and storing a value of "Yes" if `DepDelay > 0` (i.e., the flight was delayed by 1 or more minutes) and "No" if `DepDelay <= 0` (i.e., if the flight departed on time or earlier than the scheduled departure time) 

`group_by(Carrier)` is grouping by Carrier and then counting with `mutate(nflights = n())` how many flights there were per Carrier and storing this sum in a new column called `nflights`

`group_by(Carrier, late)` then regroups the data, this time by Carrier and if the flight was late or not


`mutate(nlate = n(), pct_late = (nlate / nflights) * 100)` is then creating two new columns, `nlate` -- the number of flights per late values of "Yes" and "No", respectively, and then `pct_late` -- the percent of flights per carrier that were late.  

Now, we only want the flights that were late so let us apply `select()` to keep just a few columns and then we use `filter()` to keep only rows corresponding to `late = "Yes"`. 

This will still leave us with duplicate rows but we can drop these duplicate rows via a new command, `distinct()`  -- which if left empty inside the parentheses `()` looks for all unique rows of data (each row ends up with a unique combination of all columns' values). If you want unique values of specific columns then those column names can be inserted inside the parentheses `()`

In [45]:
df1 %>%
  filter(late == "Yes") %>%
  ungroup() %>%
  select(Carrier, pct_late) %>%
  distinct() %>%
  arrange(pct_late)

Carrier pct_late
1 UA      23.96857
2 DL      27.72677
3 AA      29.86379
4 WN      41.06911

So! 24% of UA flights were late, the lowest in this group. 

What if we wanted to do this for all airlines, and we want the calculations to be done by Month as well?

In [46]:
cmhflights %>%
  select(c(Carrier, Month, DepDelay)) %>%
#  filter(Carrier %in% c("AA", "DL", "UA", "WN")) %>%
  mutate(late = case_when(
    DepDelay > 0 ~ "Yes",
    DepDelay <= 0 ~ "No"
      )
    ) %>% 
  group_by(Carrier, Month) %>%
  mutate(nflights = n()) %>%
  group_by(Carrier, Month, late) %>%
  mutate(
    nlate = n(),
    pct_late = (nlate / nflights) * 100) %>%
  filter(late == "Yes") %>%
  ungroup() %>%  
  select(Carrier, Month, pct_late) %>%
  distinct() %>%
  arrange(pct_late)

Carrier Month pct_late
1  OO      8     11.60714
2  OO      9     14.15094
3  OO      2     16.00000
4  UA      9     17.85714
5  OO      3     18.10345
6  UA      2     18.58974
7  DL      9     20.19704
8  UA      5     20.22472
9  UA      4     20.42254
10 OO      4     21.18056
11 AA      9     21.40921
12 DL      2     21.52642
13 DL      8     22.52252
14 AA      2     23.05556
15 UA      3     23.31606
16 UA      8     24.15730
17 EV      9     24.82993
18 DL      3     25.23020
19 OO      5     25.37879
20 OO      7     25.94937
21 EV      2     25.96685
22 AA      8     26.63934
23 OO      6     26.66667
24 EV      8     27.87356
25 F9      4     28.12500
26 AA      4     28.28054
27 F9      9     28.33333
28 WN      2     29.13870
29 UA      6     29.34783
30 EV      4     30.06329
⋮  ⋮       ⋮     ⋮       
34 AA      3     30.86124
35 DL      5     30.97983
36 EV      6     31.19777
37 EV      5     31.56028
38 AA      5     32.35294
39 UA      7     32.47423
40 DL      6     32.97214
41 EV      3     33.57843
42 AA      6     33.60825
43 DL      7     33.67953
44 WN      1     34.03194
45 OO      1     34.14634
46 AA      1     34.36693
47 F9      8     34.94624
48 AA      7     35.40000
49 WN      9     36.20406
50 F9      3     36.47059
51 EV      7     36.65339
52 WN      4     36.68342
53 EV      1     36.69725
54 F9      2     36.76471
55 F9      5     37.09677
56 F9      6     37.22222
57 WN      3     38.39325
58 F9      7     40.32258
59 WN      5     44.20078
60 WN      8     48.96485
61 WN      6     49.66730
62 WN      7     49.81601
63 F9      1     51.31579

Before we move on, I want to point out something about `case_when()`. Specifically, we used it to create a new column called `late` from numeric values found in `DepDelay`. But what if we wanted to create a new column from a column that had categorical variables in it, like `Dest` or `Carrier`? Easy.

In [47]:
cmhflights %>%
    mutate(
      carrier_name = case_when(
          Carrier == "AA" ~ "American Airlines",
          Carrier == "DL" ~ "Delta Airlines",
          Carrier == "UA" ~ "United Airlines",
          Carrier == "EV" ~ "Express Jet",
          Carrier == "F9" ~ "Frontier Airlines",
          Carrier == "WN" ~ "Southwest Airlines",
          Carrier == "OO" ~ "SkyWest Airlines"
      )
  ) %>%
    select(Carrier, carrier_name) %>%
    group_by(Carrier, carrier_name) %>%
    tally()

Carrier carrier_name       n    
1 AA      American Airlines   3891
2 DL      Delta Airlines      5446
3 EV      Express Jet         3056
4 F9      Frontier Airlines   1568
5 OO      SkyWest Airlines    2041
6 UA      United Airlines     1527
7 WN      Southwest Airlines 18464

Second, `case_when()` includes an option that cuts down on our work. In particular, say I want to create a new column and label its values as "Weekend" if the DayOfWeek is Saturday or Sunday and "Weekday" if DayOfWeek is any other day. In doing this, it would serve us well to remember that the week begins on Sunday so DayOfWeek == 1 is Sunday, not Monday. 

In [48]:
cmhflights %>%
  mutate(
      weekend = case_when(
          DayOfWeek %in% c(7, 1) ~ "Yes",
          TRUE ~ "No"
      )
    ) %>%
  select(DayOfWeek, weekend) %>%
  distinct()

DayOfWeek weekend
1 7         Yes    
2 1         Yes    
3 2         No     
4 3         No     
5 4         No     
6 5         No     
7 6         No

Notice how `TRUE` swept up all other values of `DayOfWeek` and coded them as "No."  

One final showcasing of `case_when()`. In **Module 01** we looked at the `hsb2` data and created some `factors` for columns such as female, ses, schtyp, and so on. Well, let us see how the same thing could be done with `case_when()`.

In [49]:
read.table(
  'https://stats.idre.ucla.edu/stat/data/hsb2.csv',
  header = TRUE,
  sep = ","
  ) -> hsb2

In [50]:
hsb2 %>%
  mutate(
    female.f = case_when(
      female == 0 ~ "Male",
      female == 1 ~ "Female"),
    race.f = case_when(
      race == 1 ~ "Hispanic",
      race == 2 ~ "Asian",
      race == 3 ~ "African-American",
      TRUE ~ "White"),
    ses.f = case_when(
      ses == 1 ~ "Low",
      ses == 2 ~ "Medium",
      TRUE ~ "High"),
    schtyp.f = case_when(
      schtyp == 1 ~ "Public",
      TRUE ~ "Private"),
    prog.f = case_when(
      prog == 1 ~ "General",
      prog == 2 ~ "Academic",
      TRUE ~ "Vocational")
    ) -> hsb2

hsb2

id  female race ses schtyp prog read write math science socst female.f
1    70 0      4    1   1      1    57   52    41   47      57    Male    
2   121 1      4    2   1      3    68   59    53   63      61    Female  
3    86 0      4    3   1      1    44   33    54   58      31    Male    
4   141 0      4    3   1      3    63   44    47   53      56    Male    
5   172 0      4    2   1      2    47   52    57   53      61    Male    
6   113 0      4    2   1      2    44   52    51   63      61    Male    
7    50 0      3    2   1      1    50   59    42   53      61    Male    
8    11 0      1    2   1      2    34   46    45   39      36    Male    
9    84 0      4    2   1      1    63   57    54   58      51    Male    
10   48 0      3    2   1      2    57   55    52   50      51    Male    
11   75 0      4    2   1      3    60   46    51   53      61    Male    
12   60 0      4    2   1      2    57   65    51   63      61    Male    
13   95 0      4    3   1      2    73   60    71   61      71    Male    
14  104 0      4    3   1      2    54   63    57   55      46    Male    
15   38 0      3    1   1      2    45   57    50   31      56    Male    
16  115 0      4    1   1      1    42   49    43   50      56    Male    
17   76 0      4    3   1      2    47   52    51   50      56    Male    
18  195 0      4    2   2      1    57   57    60   58      56    Male    
19  114 0      4    3   1      2    68   65    62   55      61    Male    
20   85 0      4    2   1      1    55   39    57   53      46    Male    
21  167 0      4    2   1      1    63   49    35   66      41    Male    
22  143 0      4    2   1      3    63   63    75   72      66    Male    
23   41 0      3    2   1      2    50   40    45   55      56    Male    
24   20 0      1    3   1      2    60   52    57   61      61    Male    
25   12 0      1    2   1      3    37   44    45   39      46    Male    
26   53 0      3    2   1      3    34   37    46   39      31    Male    
27  154 0      4    3   1      2    65   65    66   61      66    Male    
28  178 0      4    2   2      3    47   57    57   58      46    Male    
29  196 0      4    3   2      2    44   38    49   39      46    Male    
30   29 0      2    1   1      1    52   44    49   55      41    Male    
⋮   ⋮   ⋮      ⋮    ⋮   ⋮      ⋮    ⋮    ⋮     ⋮    ⋮       ⋮     ⋮       
171 119 1      4    1   1      1    42   57    45   50      43    Female  
172  13 1      1    2   1      3    47   46    39   47      61    Female  
173  51 1      3    3   1      1    42   36    42   31      39    Female  
174  26 1      2    3   1      2    60   59    62   61      51    Female  
175  36 1      3    1   1      1    44   49    44   35      51    Female  
176 135 1      4    1   1      2    63   60    65   54      66    Female  
177  59 1      4    2   1      2    65   67    63   55      71    Female  
178  78 1      4    2   1      2    39   54    54   53      41    Female  
179  64 1      4    3   1      3    50   52    45   58      36    Female  
180  63 1      4    1   1      1    52   65    60   56      51    Female  
181  79 1      4    2   1      2    60   62    49   50      51    Female  
182 193 1      4    2   2      2    44   49    48   39      51    Female  
183  92 1      4    3   1      1    52   67    57   63      61    Female  
184 160 1      4    2   1      2    55   65    55   50      61    Female  
185  32 1      2    3   1      3    50   67    66   66      56    Female  
186  23 1      2    1   1      2    65   65    64   58      71    Female  
187 158 1      4    2   1      1    52   54    55   53      51    Female  
188  25 1      2    2   1      1    47   44    42   42      36    Female  
189 188 1      4    3   2      2    63   62    56   55      61    Female  
190  52 1      3    1   1      2    50   46    53   53      66    Female  
191 124 1      4    1   1      3    42   54    41   42      41    Female  
192 175 1      4    3   2    

-----------

## Some other `dplyr()` commands
We have seen `count()` in action but let us see it again, in a slightly different light. In particular, say I want to know how many unique destinations are there connected by air from Columbus. 

### count()

In [51]:
cmhflights %>%
  filter(Origin == "CMH") %>%
  count(Dest, sort = TRUE)

Dest n   
1  ATL  2884
2  MDW  1511
3  MCO  1148
4  DFW  1122
5  DEN   971
6  BWI   948
7  LAS   815
8  PHX   815
9  ORD   803
10 EWR   736
11 MSP   677
12 TPA   595
13 BNA   520
14 RSW   509
15 DCA   507
16 STL   495
17 DTW   485
18 LAX   483
19 BOS   475
20 IAH   460
21 DAL   273
22 OAK   273
23 FLL   240
24 LGA   173
25 PHL    57
26 MSY    22

Note: There is no need for `group_by()` here. And `sort = TRUE` arranges the result in descending order of the frequency (`n`). Here is another code example, this time adding Carrier to the mix. 

In [52]:
cmhflights %>%
  filter(Origin == "CMH") %>%
  count(Carrier, Dest, sort = TRUE)

Carrier Dest n   
1  DL      ATL  2141
2  WN      MDW  1511
3  AA      DFW  1122
4  WN      BWI   948
5  WN      MCO   929
6  WN      ATL   743
7  EV      EWR   736
8  WN      TPA   595
9  UA      ORD   577
10 WN      LAS   542
11 AA      PHX   526
12 WN      BNA   520
13 WN      DCA   507
14 WN      STL   495
15 WN      RSW   486
16 WN      BOS   475
17 EV      IAH   458
18 OO      DTW   428
19 WN      DEN   384
20 DL      MSP   336
21 WN      PHX   289
22 F9      DEN   273
23 F9      LAS   273
24 WN      DAL   273
25 WN      OAK   273
26 OO      MSP   264
27 DL      LAX   242
28 AA      LAX   241
29 WN      FLL   240
30 F9      MCO   217
31 OO      ORD   199
32 UA      DEN   185
33 EV      LGA   173
34 OO      DEN   129
35 EV      MSP    77
36 AA      PHL    57
37 EV      DTW    55
38 EV      ORD    27
39 WN      MSY    22
40 F9      RSW    21
41 DL      DTW     2
42 DL      MCO     2
43 EV      RSW     2
44 OO      IAH     1
45 UA      IAH     1

How does this help us? Well, now we know that if we were flying to Atlanta, Delta would have the most flights, but if we were flying to the Chicago area then Southwest should be our pick.  

### n_distinct() 
Another useful command is `n_distinct()`, useful in the sense of allowing us to calculate the the number of distinct values of any column. For example, say I want to know how many unique aircraft (not airlines) are there in this data-set. 

In [53]:
cmhflights %>%
  summarise(n_distinct(TailNum))

n_distinct(TailNum)
1 2248

### top_n()
If you want to see the top 'n' number of observations, for example the 4 airlines with the most aircraft, you can lean on `top_n()`, as shown below.

In [54]:
cmhflights %>%
  group_by(Carrier) %>%
  summarise(num.flights = n_distinct(TailNum)) %>%
  arrange(-num.flights) %>% 
  top_n(4)

Selecting by num.flights


Carrier num.flights
1 WN      751        
2 DL      539        
3 UA      289        
4 OO      222

I am also curious about which aircraft has flown the most, and then maybe 9 other aircraft that follow in descending order.

In [55]:
cmhflights %>%
  filter(!is.na(TailNum)) %>% # Removing some missing cases 
  group_by(TailNum) %>%
  tally() %>% 
  arrange(-n) %>%
  top_n(4)

Selecting by n


TailNum n 
1 N396SW  74
2 N601WN  66
3 N646SW  64
4 N635SW  62

### join()
You will, from time to time, need to merge multiple data-sets together. For example, say I have the following data-sets I have created for demonstration purposes. 

In [56]:
tibble(
  Name = c("Tim", "Tammy", "Bubbles", "Panda"),
  Score = c(5, 8, 9, 10)
    ) -> df1

In [57]:
tibble(
  Name = c("Tim", "Tammy", "Bubbles"),
  Age = c(25, 78, 19)
    ) -> df2

In [58]:
tibble(
  Name = c("Tim", "Tammy", "Panda"),
  Education = c("BA", "PhD", "JD")
    ) -> df3

In [59]:
df1; df2; df3

Name    Score
1 Tim      5   
2 Tammy    8   
3 Bubbles  9   
4 Panda   10

Name    Age
1 Tim     25 
2 Tammy   78 
3 Bubbles 19

Name  Education
1 Tim   BA       
2 Tammy PhD      
3 Panda JD

Notice that Panda is absent from `df2` and Bubbles is absent from `df3`. So if we wanted to build ONE data-set with all data for Tim, Tammy, Bubbles, and Panda, some of the information would be missing for some of these folks. But how could we construct ONE data-set? Via one of a few `join()` commands. 

#### full_join() 
Let us start with a simple full_join, where we link up every individual in df1 or df2 or df3 **regardless of whether they are seen in both data-sets**. 

In [60]:
df1 %>%
  full_join(df2, by = "Name") %>%
  full_join(df3, by = "Name") 

Name    Score Age Education
1 Tim      5    25  BA       
2 Tammy    8    78  PhD      
3 Bubbles  9    19  NA       
4 Panda   10    NA  JD

Pay attention to two things: (i) Name connects the records in each data-set, and so it must be spelled exactly the same for a link to be made, and (ii) the `full_join()` links up all individuals regardless of whether they are missing any information in any of the data-sets. This is usually how most folks will link up multiple files unless they only want records found in a master file. For example, say I want to link up df2 and df3 but only such that the final result will include all records found in BOTH df2 and df3, with df2 serving as the master data-set. Eh?

In [61]:
df2 %>%
  left_join(df3, by = "Name")  

Name    Age Education
1 Tim     25  BA       
2 Tammy   78  PhD      
3 Bubbles 19  NA

Notice that Panda is dropped because it is not found in df2. 

Maybe you want df3 to be the master file, in which case you would see a different result (with Bubbles not seen in the result since Bubbles is found in df2 but not in df3): 

In [62]:
df3 %>%
  left_join(df2, by = "Name")  

Name  Education Age
1 Tim   BA        25 
2 Tammy PhD       78 
3 Panda JD        NA

Rarely, but definitely not "never," you may want to see the records that are not found in both. Here, anti_join() comes in handy, thus:

In [63]:
df2 %>%
  anti_join(df3, by = "Name")

Name    Age
1 Bubbles 19

In [64]:
df3 %>%
  anti_join(df2, by = "Name")

Name  Education
1 Panda JD

--------------------

## Two other useful commands 
### {santoku}
Every now and then you may want to or need to create a grouped version of some numeric variable. For example, we have DepDelay for all flights but want to group this into `quartiles`. How can we do that? In many ways but the easiest might be to use a specific library -- `{santoku}`. Say, for example, I want to create 4 groups of `dep_delay`, and I want these such that we are grouping `DepDelay` into the bottom 25%, next 25%, the next 25%, and finally the highest 25%. Wait, these are the `quartiles`! Fair enough, but how can I do this? 

In [70]:
install.packages("santoku")

Installing package into ‘/Users/ruhil/Library/R/arm64/4.3/library’
(as ‘lib’ is unspecified)




The downloaded binary packages are in
	/var/folders/qh/6q39v0755_54rxmbl8m5ttnwy0twd7/T//RtmpHYaGP5/downloaded_packages


In [71]:
library(santoku)

cmhflights %>%
  mutate(
    depdelay_groups = chop_equally(DepDelay, groups = 4)
      ) %>%
  group_by(depdelay_groups) %>%
  tally()


Attaching package: ‘santoku’


The following object is masked from ‘package:tidyr’:

    chop




depdelay_groups n    
1 [-27, -5)        6887
2 [-5, -2)         9267
3 [-2, 7)         10143
4 [7, 1323]        9225
5 NA                471

What if we wanted to slice up DepDelay in specific intervals, first at 0, then at 15, then at 30, and then at 45? 

In [72]:
cmhflights %>%
  mutate(
    depdelay_groups = chop(DepDelay, breaks = c(0, 15, 30, 45))
      ) %>%
  group_by(depdelay_groups) %>%
  tally()

depdelay_groups n    
1 [-27, 0)        21108
2 [0, 15)          7883
3 [15, 30)         2567
4 [30, 45)         1240
5 [45, 1323]       2724
6 NA                471

We could also create quintiles (5 groups) or deciles (10 groups) as shown below: 

In [73]:
cmhflights %>%
  filter(!is.na(DepDelay)) %>%
  mutate(
    depdelay_groups = chop_quantiles(
      DepDelay, c(0.2, 0.4, 0.6, 0.8)
    )
  ) %>%
  group_by(depdelay_groups) %>%
  tally()

depdelay_groups n   
1 [0%, 20%)       6887
2 [20%, 40%)      6342
3 [40%, 60%)      7879
4 [60%, 80%)      7035
5 [80%, 100%]     7379

In [74]:
cmhflights %>%
  filter(!is.na(DepDelay)) %>%
  mutate(
    depdelay_groups = chop_quantiles(
      DepDelay, seq(0.1, 0.9, by = 0.1)
    )
  ) %>%
  group_by(depdelay_groups) %>%
  tally()

depdelay_groups n   
1  [0%, 10%)       3006
2  [10%, 20%)      3881
3  [20%, 30%)      3344
4  [30%, 40%)      2998
5  [40%, 50%)      2925
6  [50%, 60%)      4954
7  [60%, 70%)      3262
8  [70%, 80%)      3773
9  [80%, 90%)      3805
10 [90%, 100%]     3574

You could also ask for groups such that they have the same width. Below we create 4 groups. Note that the width of each group is exactly 337.5 ... 
1323 - 985.5 = 337.5; 985.5 - 648 = 337.5; and so on. In this particular example this chopping isn't very useful since we end up with almost all of the data in the very first group. However, there are occasions where equal widths are useful. 

In [75]:
cmhflights %>%
  filter(!is.na(DepDelay)) %>%
  mutate(
    depdelay_groups = chop_evenly(
        DepDelay,
        4
    )
  ) %>%
  group_by(depdelay_groups) %>%
  tally()

depdelay_groups n    
1 [-27, 310.5)    35439
2 [310.5, 648)       68
3 [648, 985.5)        5
4 [985.5, 1323]      10

### ordered()
More often than we would like to see happen, we often end up with categorical variables that should follow a certain order but do not. For example, say you have survey data where people were asked to respond whether they Agree, are Neutral, or Disagree with some statement. Let us also assume that the frequencies are as follows:

In [76]:
tibble(
  response = c(
      rep("Agree", 25), 
      rep("Neutral", 30), 
      rep("Disagree", 45)
      )
    ) -> mydf

In [77]:
mydf %>%
  group_by(response) %>%
  tally()

response n 
1 Agree    25
2 Disagree 45
3 Neutral  30

Notice how the responses are out of order, with Agree followed by Disagree, then Neutral, since R defaults to alphabetic ordering for anything that is a categorical variable. One way to ensure the correct ordering of categorical variables is via `ordered`, as shown below. 

In [78]:
mydf %>%
  mutate(
      ordered.response = ordered(
          response,
          levels = c("Agree", "Neutral", "Disagree")
      )
    ) %>%
  group_by(ordered.response) %>%
  tally()

ordered.response n 
1 Agree            25
2 Neutral          30
3 Disagree         45

## Concluding thoughts 
We have a covered a lot of ground here but every inch has been critical space. Google any question we have tackled and you will see how many R-users ask the same questions ... how do I calculate mean for groups in R? What you have seen is the heart of the `dplyr()` package. We saw grouped operations, we saw the use of summarise, mutate, case_when, distinct, filter, arrange, select, count, and tally. I will let you in on a secret; while these are core functions, there are others you could experiment with. Look up the cheat-sheet [here](https://rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf). 

Practice what we have done, with the `{nycflights13}` data-set perhaps to get something familiar yet sufficiently different to test your fundamentals. Maybe pick a non-travel data-set altogether, perhaps one of the `tidytuesday` data-sets. What is that you ask? Discover it for yourself [here](https://github.com/rfordatascience/tidytuesday). Bon voyage! Don't go too far because we will be working with two new packages next week -- `{tidyr}` and `{lubridate}`. 

-----

# Exercises for Practice 
## Exercise 01
Why are our best and most experienced employees leaving prematurely? 

The data [available here](https://aniruhil.github.io/avsr/teaching/dataviz/HR_comma_sep.csv) includes information on several current and former employees of an anonymous organization. Fields in the data-set include: 

| Variable | Description |
| :---- | :---- |
| satisfaction_level | = Level of satisfaction (numeric; 0-1) |
| last_evaluation | = Evaluation score of the employee (numeric; 0-1) |
| number_project | = Number of projects completed while at work (numeric) |
| average_monthly_hours | = Average monthly hours spent at the workplace (numeric)  |
| time_spend_company | = Number of years spent in the company (numeric) |
| Work_accident | = Whether the employee had a workplace accident (categorical; 1 = yes or 0 = no) |
| left | = Whether the employee left the workplace or not (categorical; 1 = left or 0 = stayed)  |
| promotion_last_5years | = Whether the employee was promoted in the last five years (categorical; 1 = yes or 0 = no) |
| sales | = Department in which they work (categorical) |
| salary | = Relative level of salary (categorical; low, med, and high) |

(a) Read in the csv-format data-set, naming it `hrdata` and save it in RData format as `hrdata.RData` 

(b) Create new variables that add labels to Work_accident, left, promotion_last_5years, and add these three to `hrdata`.  

(c) Now retain only employees who left the company, and had not been promoted in the last five years. Save this result as `hr01`

(d) In this `hr01` data-set, how many employees do you have per sales department? What sales department has the most number of employees? 

(e) By sales department, calculate mean and standard deviation of (i) satisfaction_level, and (ii) last_evaluation. 

(f) What department has the lowest mean satisfaction? Which department has the highest variation in satisfaction?  

(g) Create a new variable that groups `average_montly_hours` into 4 groups. You can let the group cut-points be chosen automatically with `chop_evenly()`. Then show the frequencies of each group.

## Exercise 02
Thanks to the frenetic work of many individuals, the global spread of the Novel Coronavirus (COVID-19) has been tracked and the data made available for analysis. [Yanchang Zhao](https://rdatamining.wordpress.com/2020/03/10/coronavirus-data-analysis-with-r-tidyverse-and-ggplot2/) is one such individual and for this exercise we will use a spcific version of his data that I have named `cvdata.RData` is available in the `data` folder. Read it in via the `load()` command. We can then answer a few questions. Note the contents: 

+ `country =` name of the country 
+ `date =` date of incidents as recorded 
+ `confirmed =` cumulative count of the number of people who tested positive  
+ `deaths =` cumulative count of the number of people lost to Covid-19 
+ `deaths =` cumulative count of the number of people recovered  

(a) Filter the data-set so that we have only one row per country, the data from March 10, 2020 and call it `cv0310`. 

(b) How many countries have lost `at least one` person to this tragedy? "Others" should not show up as one of the countries.  

(c) What 10 countries have had the most number of confirmed cases? "Others" should not show up as one of the countries. Also ensure the result is organized in descending order of the number of confirmed cases. 

(d) Calculate the `fatality_rate`, defined for our purposes as the percent of deaths. excluding "Others", and only keeping countries that have had `at least 10` confirmed cases, arrange the result to show the top-10 countries in descending order of `fatality_rate`.  

(e) Say we only want to focus on the Baltic countries (Estonia, Latvia, and Lithuania) as a unified group and compare this group to the ASEAN nations (Brunei, Cambodia, Indonesia, Laos, Malaysia, Myanmar, Philippines, Singapore, Thailand, and Vietnam). Use `cv0310` to complete the followng tasks: 

(i) Create a new variable called `region` that only takes on two values -- "Baltic" if the country is a Baltic country and "Asean" if the country is an ASEAN country. 

(ii) Use this variable to calculate the total number of confirmed cases in each region. 